src/entry.c

```diff
@@ -39,8 +41,18 @@
 static void xcb_xrm_append_char(xcb_xrm_entry_t *entry, xcb_xrm_entry_parser_state_t *state,
         const char str) {
     if (state->buffer_pos == NULL) {
-        memset(&(state->buffer), 0, sizeof(state->buffer));
+        FREE(state->buffer);
+        state->buffer = calloc(1, BUFFER_SIZE);
         state->buffer_pos = state->buffer;
+        if (state->buffer == NULL) {
+            /* Let's ignore this character and try again next time. */
+            return;
+        }
+    }
+
+    /* Increase the buffer if necessary. */
+    if ((state->buffer_pos - state->buffer) % (BUFFER_SIZE + 1) == BUFFER_SIZE) {
```

Comment:
Does this actually work? Let's check:
- Initially `buffer_pos - buffer == 0` and `0 % (BUFFER_SIZE + 1) != BUFFER_SIZE`, which is fine
- After `BUFFER_SIZE-1` characters `buffer_pos - buffer == BUFFER_SIZE - 1` and `(BUFFER_SIZE - 1) % (BUFFER_SIZE + 1) != BUFFER_SIZE`, which is still fine, buffer is full after the next character (However, this is bad, see below)
- For the next character, the buffer is resized.
- After another `BUFFER_SIZE` characters: `buffer_pos - buffer == 2*BUFFER_SIZE = (BUFFER_SIZE + 1) + (BUFFER_SIZE - 1)`. Modulo `BUFFER_SIZE + 1` this gives the wrong value, so the buffer is resized one byte too late.

Now comes the however mentioned above:
`xcb_xrm_finalize_component` appends a `\0`-byte to the buffer. With the above code, this will already be a writer beyond the bound of the buffer when the value has exactly length `BUFFER_SIZE`.

Thus my suggested fix is to replace the code here with `(state->buffer_pos - state->buffer) % BUFFER_SIZE == BUFFER_SIZE - 1`. That way it should reallocate the buffer whenever appending the next character would cause an overflow.